Add support for DPPI on nRF54H20

anangl opened the pull request:
- add dts nodes and corresponding bindings for DPPIC and IPCT peripherals on nRF54H20 EngA
- provide the nrfx configuration required by the nrfx_gppi_dppi_ppib helper and include the helper into compilation when DPPIC nodes are enabled in dts

anangl committed:
dts: nordic: Add DPPIC and IPCT nodes to nRF54H20 EngA

And add the corresponding bindings.

Move alse the already existing bindings for nrf-dppic, nrf-ppi,
and nrf-ipc so they are located together with the new ones and
in more appropriate folders (DPPIC and PPI peripherals are not
related to ARM, and IPC is for sending and receiving events,
not messages, so ipm/ does not seem to be the best fit for it).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
modules: hal_nordic: Enable support for DPPI on nRF54H20

Include into compilation the nrfx_gppi_dppi_ppib helper and related
interconnect layers when DPPIC nodes are enabled in DTS. Provide macro
definitions required by those interconnect layers based on information
from devicetree (the nrf_grtc_timer is only modified because a macro
that it uses became more generic).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
manifest: Update hal_nordic revision

Pull in fixes in the nrfx_gppi_dppi_ppib helper so that it can be
compiled successfuly.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
cc @adamkondraciuk

anangl commented:
Rebased.

gmarull commented on dts/bindings/ipc/nordic,nrf-ipct-common.yaml:
```diff
@@ -0,0 +1,43 @@
+      in the array, therefore requiring the length of this property to be
+      a factor of 3.
+
+      For example, if channel 2 is to be mapped to Secure Domain (ID: 1) IPCT
+      channel 4, then the array "unit" would be <2 1 4>.
```
note: I'd provide definitions for domain identifiers, otherwise reading dt files becomes a bit cumbersome, specially if future SoCs use other ids. These could be defined in the main/root DT file itself (no need to include anything)

anangl replied:
Added. Please take a look.

gmarull commented on dts/bindings/ipc/nordic,nrf-ipct-global.yaml:
```diff
@@ -0,0 +1,17 @@
+include: ["nordic,nrf-ipct-common.yaml", "nordic,split-channels.yaml"]
+
+properties:
+  global-domain-id:
+    required: true
+    type: int
+    description: |
+      Global IPCT instances reside on specific buses within the Global Domain,
+      such as fast and slow, which have different IDs that do not match the
+      standard Global Domain ID presented in their address.
```
question: what is the real purpose of this property? Feels like the `domain` one that existed for mbox-nrf-ids

anangl replied:
I'm not sure. I saw that it is used by regtool so I thought it should be kept. @57300 Could you comment on that?

gmarull replied:
can domain be inferred by other means, like it happened with ipc related stuff? I'd like to make sure we do not add redundancy in DT just for a tool.

57300 replied:
What is redundant here? This can't be inferred from the address, as the description says, because our brilliant hardware designers assigned both IPCT instances to domain ID 15, instead of 12 and 13.

The IPC stuff we solved by parsing node labels, which I still don't think is great. Do you want to do the same here, and label the nodes `globalfast_ipct` and `globalslow_ipct`?

gmarull replied:
I see your point. To me, both options are equally bad (nodelabels adding info, or, adding properties not consumed by driver code). Sad that HW has been laid out like this. Let's take the shortest path here, we can always address this later.

gmarull commented on dts/bindings/ipc/nordic,nrf-ipct-common.yaml:
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: Nordic IPCT (Interprocessor Communication Transceiver)
```
nit: add example snippets, these bindings are complex enough that it would be good to have, also with some extended details on what IPCT does

anangl committed:
soc: nordic_nrf: Add address validation entries for nRF54H/L dts nodes

Add entries for checking if the recently added dts nodes for nRF54H20
and nRF54L15 have correct base addresses (if they match those provided
by MDK).

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: nordic: Add missing ECB020 node in nRF54H20 EngA

This is a follow-up to commit 4db40601dd7f0b98a48f3e6072bf44643a4ea3db.

There are two ECB instances in nRF54H20 and the above commit added
only one of them.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -756,27 +757,127 @@
+ * - NRFX_INTERCONNECT_APB_LOCAL_DPPI_DEFINE
+ * based on information from devicetree.
+ */
+#if	defined(CONFIG_DT_HAS_NORDIC_NRF_DPPIC_GLOBAL_ENABLED) || \
+	defined(CONFIG_DT_HAS_NORDIC_NRF_DPPIC_LOCAL_ENABLED)
```
That's actually the same. But yes, `DT_HAS_COMPAT_STATUS_OKAY()` will be probably more readable. I'll make this change.

anangl replied:
done
